rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,251 @@
+
+The following code shows a convolution kernel implementation using the stream handle. The streams are created during the pluggable device creation. The placer decides which device to use for each OP in the graph. Then the streams associated with the device are used to construct the OpKernelContext for the op computation during the graph execution.
+```cpp
+void Conv_Compute(TF_OpKernelContext*) {
+  TF_GetInput(context, input_index, &input, &status);
+  TF_GetInput(context, filter_index, &filter, &status);
+  auto output = TF_AllocateOutput(context, output_index, TF_Float32, dims, num_dims, len, status);
+  SE_Stream se_stream = TF_GetStream(TF_OpKernelContext);
+  auto native_stream = static_cast<native_stream_type>(se_stream->stream_handle);
+  my_conv_impl(input, filter, output, native_stream);
+}
+```
+Kernel and op registration and implementation API [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md) needs to be extended to retrieve streams/device context from the TF_OpKernelContext, besides inputs and outputs. 
+
+### **Alternatives Considered**
+
+* Without this RFC, end users need to change the python code to import the third-party device plugin. 
+
+* Without this RFC, the third-party device vendor may implement the LocalDevice interface, which is not a C API interface and may interact with potential C++ ABI incompatibility issues.  
+
+### **Performance Implications**
+
+* We don’t expect performance impact due to this RFC. The functions described by this RFC are realized at the initialization stage. 
+
+### **Dependencies**
+
+* This RFC doesn’t add new dependencies to external libraries. 
+
+* It depends on three modular TensorFlow related RFC 
+
+    * Modular TensorFlow  [RFC](https://github.com/tensorflow/community/pull/77)
+
+    * StreamExecutor C interface [RFC](https://github.com/tensorflow/community/pull/257)
+
+    * Kernel and op registration and implementation API [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md) 
+
+### **Engineering Impact**
+
+* The impact to binary size / startup time / build time / test times are minimum. 
+
+* The TensorFlow team will maintain this code. 
+
+### **Platforms and Environments**
+
+* The pluggable device mechanism is based on `LoadLibrary()` so should work on all the platforms supported by `LoadLibrary`. The other enhancement to tensorflow proper is platform independent.
+
+### **Best Practices**
+
+* This works with Modular TensorFlow which will be the only way to integrate new third-party devices to the current TensorFlow stack. 
+
+### **Compatibility**
+
+The RFC promotes the current TensorFlow ecosystem as it supports plugging new devices to TensorFlow.  
+
+We don't expect this proposal to impact with other parts of the TensorFlow ecosystem. It doesn't support TFLite. It should not impede distribution strategies and would not interact with tf.fuction and SaveModel.  
+
```

Comment:
This needs to show how the pluggable device is available from python. Do we expect the user to call a tf.config.init_pluggable_device(path_to_so_file)? Or do we expect it to be via environment variables? And what device strings are going to be used to access the pluggable device with tf.device?

Reply:
We expect user don't need to load the library manually, tensorflow will automatically load the library from the specific directory, such as `...python_dir.../site-packages/tensorflow-plugins`.[Modular Tensorflow Plugin load](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md#how-plugins-are-loaded)
 The device strings for tf.device() is registered from plugin, it can be "GPU" name with higher priority, and user can run it with legacy gpu model. or it can be another device name, user need to list the device first(like tf.config.list_physical_device) and got the name for tf.device().